pkg/dogstatsd/listeners/named_pipe_windows.go

```diff
@@ -0,0 +1,143 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-2020 Datadog, Inc.
+// +build windows
+
+package listeners
+
+import (
+	"io"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/config"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+
+	"github.com/Microsoft/go-winio"
+)
+
+var namedPipeTelemetry = newListenerTelemetry("named_pipe", "Named Pipe")
+
+const pipeNamePrefix = `\\.\pipe\`
+
+// NamedPipeListener implements the StatsdListener interface for named pipe protocol.
+// It listens to a given pipe name and sends back packets ready to be processed.
+// Origin detection is not implemented for named pipe.
+type NamedPipeListener struct {
+	pipe          net.Listener
+	packetManager *packetManager
+	connections   map[net.Conn]struct{}
+	mux           sync.Mutex
+}
+
+// NewNamedPipeListener returns an named pipe Statsd listener
+func NewNamedPipeListener(pipeName string, packetOut chan Packets, sharedPacketPool *PacketPool) (*NamedPipeListener, error) {
+	bufferSize := config.Datadog.GetInt("dogstatsd_buffer_size")
+	return newNamedPipeListener(
+		pipeName,
+		bufferSize,
+		newPacketManagerFromConfig(
+			packetOut,
+			sharedPacketPool))
+}
+
+func newNamedPipeListener(
+	pipeName string,
+	bufferSize int,
+	packetManager *packetManager) (*NamedPipeListener, error) {
+
+	config := winio.PipeConfig{
+		InputBufferSize:  int32(bufferSize),
+		OutputBufferSize: 0,
+	}
+	pipePath := pipeNamePrefix + pipeName
+	pipe, err := winio.ListenPipe(pipePath, &config)
+
+	if err != nil {
+		return nil, err
+	}
+
+	listener := &NamedPipeListener{
+		pipe:          pipe,
+		packetManager: packetManager,
+		connections:   make(map[net.Conn]struct{}),
+	}
+
+	log.Debugf("dogstatsd-named-pipe: %s successfully initialized", pipe.Addr())
+	return listener, nil
+}
+
+// Listen runs the intake loop. Should be called in its own goroutine
+func (l *NamedPipeListener) Listen() {
+	for {
+		conn, err := l.pipe.Accept()
+		switch {
+		case err == nil:
+			l.mux.Lock()
+			l.connections[conn] = struct{}{}
+			l.mux.Unlock()
+			go l.listenConnection(conn)
```

Comment:
Did you have the chance to valide that spawning so many routines isn't a performance issue? I mean, spawning that many routines may not be an issue itself but every routines are competiting for the `l.mux` mutex, and in cases where we receive hundreds of thousands of metrics per seconds, I don't see how this could not be an issue?

Reply:
This code is called when a new client connects to the server. It happens once by client.

Reply:
Ah OK I thought it was each time we received a packet, cool!